File: Chapter08/Java/ChallengeResponseClient/app/src/main/java/net/zenconsult/android/ChallengeResponseClientActivity.java
package net.zenconsult.android;

import android.app.Activity;
import android.os.Bundle;
import android.widget.Button;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

public class ChallengeResponseClientActivity extends AppCompatActivity {
    /** Called when the activity is first created. */
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.main);
        final Activity activity = this;

        final Button button = (Button) findViewById(R.id.button1);
        button.setOnClickListener((v) -> {
            new Thread(() -> {
                Comms c = new Comms(activity);
                String challenge = c.getChallenge();
                CRAM cram = new CRAM(activity);
                String hash = cram.generate(challenge);
                String reply = c.sendResponse(hash);
                runOnUiThread(() -> {
                    if (c.authorized(reply)) {
                        Toast toast = Toast.makeText(
                                activity.getApplicationContext(), "Login success",
                                Toast.LENGTH_LONG);
                        toast.show();
                    } else {
                        Toast toast = Toast.makeText(
                                activity.getApplicationContext(), "Login failed",
                                Toast.LENGTH_LONG);
                        toast.show();
                    }
                });
            }).start();
        });
    }
}

File: Chapter08/Kotlin/ChallengeResponseClient/app/src/main/java/net/zenconsult/android/ChallengeResponseClientActivity.kt
package net.zenconsult.android

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Button
import android.widget.Toast

class ChallengeResponseClientActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.main)

        val button = findViewById<Button>(R.id.button1)
        button.setOnClickListener {
            Thread {
                val c = Comms(this)
                val challenge = c.getChallenge()
                val cram = CRAM(this)
                val hash = cram.generate(challenge)
                val reply = c.sendResponse(hash)
                runOnUiThread {
                    if (c.authorized(reply)) {
                        Toast.makeText(this, "Login success", Toast.LENGTH_SHORT).show()
                    } else {
                        Toast.makeText(this, "Login failed", Toast.LENGTH_SHORT).show()
                    }
                }
            }.start()
        }
    }
}